src/algebra/ordered_monoid_lemmas.lean

```diff
@@ -724,3 +733,63 @@ lemma strict_mono.mul_monotone' (hf : strict_mono f) (hg : monotone g) :
 λ x y h, mul_lt_mul_of_lt_of_le (hf h) (hg h.le)
 
 end strict_mono
+
+/--
+An element `a : α` is `mul_le_cancellable` if `x ↦ a * x` is order-reflecting.
+We will make a separate version of many lemmas that require `[contravariant_class α α (*) (≤)]` with
+`mul_le_cancellable` assumptions instead. These lemmas can then be instantiated to specific types,
+like `ennreal`, where we can replace the assumption `add_le_cancellable x` by `x ≠ ∞`.
+-/
+@[to_additive /-" An element `a : α` is `add_le_cancellable` if `x ↦ a + x` is order-reflecting.
+We will make a separate version of many lemmas that require `[contravariant_class α α (+) (≤)]` with
+`mul_le_cancellable` assumptions instead. These lemmas can then be instantiated to specific types,
+like `ennreal`, where we can replace the assumption `add_le_cancellable x` by `x ≠ ∞`. "-/
+]
+def mul_le_cancellable [has_mul α] [has_le α] (a : α) : Prop :=
+∀ ⦃b c⦄, a * b ≤ a * c → b ≤ c
```

Comment:
Except for the implicitness on `b c`, this definition would be `contravariant α α (*) (≤)`specialized to `m = a`, right?  Why does Lean not accept `contravariant α α (*) (≤) a`?

Reply:
Yes it is.

`contravariant α α (*) (≤)` is a proposition that doesn't take other arguments, so applying it to `a` doesn't make sense.

Reply:
This suggests to me that `contravariant α α (*) (≤)` could be defined in terms of `∀ a. mul_le_cancellable` or something

Reply:
I don't think that matters...